From 6be3610ced3895be7441c936894cecf451d9486a Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Wed, 15 Nov 2023 17:48:51 +0100
Subject: [PATCH 109/217] net: dovetail: add post-routing hook for IPv4 packets

Add the ip_learn_oob_route() (in-band) hook which is called when a new
route is resolved for an egress IPv4 packet going through an
oob-enable device. This allows a companion core to cache the routing
information determined by the in-band netstack.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 net/ipv4/route.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/net/ipv4/route.c b/net/ipv4/route.c
index 261ddb654..faaffd3d3 100644
--- a/net/ipv4/route.c
+++ b/net/ipv4/route.c
@@ -2879,6 +2879,14 @@ struct dst_entry *ipv4_blackhole_route(struct net *net, struct dst_entry *dst_or
 	return rt ? &rt->dst : ERR_PTR(-ENOMEM);
 }
 
+#ifdef CONFIG_NET_OOB
+__weak
+#else
+static inline
+#endif
+void ip_learn_oob_route(struct net *net, struct flowi4 *flp4, struct rtable *rt)
+{ }
+
 struct rtable *ip_route_output_flow(struct net *net, struct flowi4 *flp4,
 				    const struct sock *sk)
 {
@@ -2894,6 +2902,14 @@ struct rtable *ip_route_output_flow(struct net *net, struct flowi4 *flp4,
 						  sk, 0));
 	}
 
+	/*
+	 * If the route goes through an oob-enabled device, the oob
+	 * core might want to learn about the routing decision, pass
+	 * it on.
+	 */
+	if (!IS_ERR(rt) && netdev_is_oob_port(rt->dst.dev))
+		ip_learn_oob_route(net, flp4, rt);
+
 	return rt;
 }
 EXPORT_SYMBOL_GPL(ip_route_output_flow);
-- 
2.30.2

